### PR TITLE
User hierarchy implementation

### PR DESCRIPTION
**Pull request by 8r1m2t0n3**

### CODE REVIEW

File: src/main/java/org/andersen/homework/MainClass.java [hunk 1]
```diff
@@ -11,30 +15,54 @@ public class MainClass {
...
+
+    System.out.println();
+
+    Admin admin = new Admin((short) RandomizerUtil.getRandomInt(0, 999));
+    System.out.println("Admin id: " + admin.getId());
+    admin.print();
+    admin.checkTicket();
+  }
+
+  public static void main(String[] args) {
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/MainClass.java]: Usually we read the code from top to bottom, so I suggest you to move main method to the top

File: src/main/java/org/andersen/homework/MainClass.java [hunk 2]
```diff
@@ -11,30 +15,54 @@ public class MainClass {
...
+
+    fullTicket.share("19374682");
+    fullTicket.share("18364529", "email@mail.com");
+  }
+
+  private static void printUsers() {
+    Client client = new Client((short) RandomizerUtil.getRandomInt(0, 999));
+    System.out.println("Client id: " + client.getId());
+    client.print();
+    client.getTicket();
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/MainClass.java]: unassigned return value? why do we need this call here?

File: src/main/java/org/andersen/homework/annotation/NullableWarning.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface NullableWarning {
+
+  String message() default "\033[3mWarning: Field [%s] is null in class [%s]!\033[0m";
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/annotation/NullableWarning.java:12]: What does \033[3m mean? ;)

File: src/main/java/org/andersen/homework/model/entity/User.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package org.andersen.homework.model.entity;
+
+import lombok.AllArgsConstructor;
+import lombok.Getter;
+import lombok.Setter;
+
+@Getter
+@Setter
+@AllArgsConstructor
+public abstract class User {
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/User.java:12]: Why have you chosen here abstract class and combine id(getting/setting) and also printRole method ?

File: src/main/java/org/andersen/homework/model/entity/Client.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+  public void printRole() {
+    System.out.println("Client");
+  }
+
+  @Override
+  public void print() {
+    System.out.println("I am a Client");
+  }
+
+  public void getTicket() {
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/Client.java]: seems like getter here - so probably the field also need here?)

File: src/main/java/org/andersen/homework/model/entity/Admin.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+  public void printRole() {
+    System.out.println("Admin");
+  }
+
+  @Override
+  public void print() {
+    System.out.println("I am an Admin");
+  }
+
+  public void checkTicket() {
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/Admin.java]: What about params here?

File: src/main/java/org/andersen/homework/model/entity/Ticket.java [hunk 7]
```diff
@@ -4,40 +4,118 @@
...
+        ", concertHallName='" + concertHallName + '\'' +
+        ", eventCode=" + eventCode +
+        ", isPromo=" + isPromo +
+        ", backpackWeightInKg=" + backpackWeightInKg +
+        ", time=" + time +
+        ", stadiumSector=" + stadiumSector +
+        '}';
+  }
+
+  private void checkFields() {
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/Ticket.java]: Does this method connected with Ticket itself? Is it an ability of a ticket? :)
What if we would like to use the annotation for other class?

File: src/main/java/org/andersen/homework/model/entity/Ticket.java [hunk 8]
```diff
@@ -4,40 +4,118 @@
...
+        ticket.priceInUsd) && Objects.equals(concertHallName, ticket.concertHallName)
+        && Objects.equals(eventCode, ticket.eventCode) && Objects.equals(isPromo,
+        ticket.isPromo) && Objects.equals(backpackWeightInKg, ticket.backpackWeightInKg)
+        && Objects.equals(time, ticket.time) && stadiumSector == ticket.stadiumSector;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(id, priceInUsd, concertHallName, eventCode, isPromo, backpackWeightInKg,
+        time,
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/Ticket.java]: formatting?)

File: src/main/java/org/andersen/homework/model/entity/Ticket.java [hunk 9]
```diff
@@ -4,40 +4,118 @@
...
-  private Short eventCode;
+  @NullableWarning
+  private final Short eventCode;
 
-  private Boolean isPromo;
+  @NullableWarning
+  private final Boolean isPromo;
 
+  @NullableWarning
+  private final Float backpackWeightInKg;
```
Comment (Darya-Drutsko) [inline src/main/java/org/andersen/homework/model/entity/Ticket.java:45]: What is the purpose of changing fields to final?)

**Review by Darya-Drutsko — CHANGES_REQUESTED** [review]
Left some comments, but overall good job!

File: src/main/java/org/andersen/homework/annotation/NullableWarning.java (hunk 3 above)
Comment (8r1m2t0n3) [reply src/main/java/org/andersen/homework/annotation/NullableWarning.java:12]: Makes font italics

File: src/main/java/org/andersen/homework/model/entity/Ticket.java (hunk 9 above)
Comment (8r1m2t0n3) [reply src/main/java/org/andersen/homework/model/entity/Ticket.java:45]: In the 3d point of the task was specified to make only "time" and "stadium sector" fields changeable

**Pull request merged**